### DISCUSSION TITLE
Design of widgets

### DISCUSSION THREAD

**Discussion by certik** [jupyter-xeus/cpp-terminal#247] [General]
The discussion started in https://github.com/jupyter-xeus/cpp-terminal/pull/239#issuecomment-1537495024. We can continue here.

**Comment by certik** [comment]
https://github.com/jupyter-xeus/cpp-terminal/pull/239#issuecomment-1539682747:

> I like Turbo Vision visual style, imitating a graphical GUI. From the programming perspective, having a `Widget` base class which can contain children and following a Qt like approach is the most familiar one to me. I did some half-baked approach on my program so far, but it would need to be more sophisticated. Should we create a new thread to talk about how to engage this development?

@iganinja, I like the Turbo Vision style as well. Married with unicode it can look really nice.

I am also familiar with the imperative Qt style. So why don't we start with that? We can always build a "react" functional style on top of this imperative style as another optional layer.

**Reply by flagarde** [reply]
@certik Concerning the utf8. Well it's a bit orthogonal but usefull, I'm first adding possibility to change the font on windows as the windows fonts are not fully BMP aware.. Seems to work I will do a PR for very preliminary feature. Maybe we could embedded the Unifont (downloaded a build time and install with cpp-terminal) to simplify tests an d good rendering in windows?

**Reply by certik** [reply]
How do you embed a font? I thought the user has to select it by hand in the terminal?

**Reply by flagarde** [reply]
@certik I mean we can put the otf file in the installed path and let user switch to it. On windows11 the default font seems ok to deal with many languages but I don't know on other version.

**Reply by flagarde** [reply]
It seems possible to inject it as bytes and call some Windows functions to load this buffer but it seems it's not working with terminal app.

**Reply by flagarde** [reply]
@certik This is what I played with : 
```cpp
 std::cout << "Running cpp-terminal version: " << Term::Version << " website : " << Term::Homepage << std::endl << std::endl;
  Term::terminal.setFont("C:\\Users\\user\\MyLovedFont.otf","Unifont");
  try
  {
    std::cout<<"Some tests : "<<std::endl;
    std::cout<<"Chinese : 中文"<<std::endl;
    std::cout<<"Arabic :  العربية"<<std::endl;
    std::cout<<"Greek : Ελληνικά"<<std::endl;
    std::cout<<"Japanese : 私の猫です。"<<std::endl;
    std::cout<<"Korean : 바게트"<<std::endl;
    std::cout<<"Russian : русский язык, rússkij âzýk"<<std::endl;
    std::cout<<"Finnish : sähköposti"<<std::endl;
    std::cout<<""<<std::endl;
    std::cout << "Press any key to quit" << std::endl;
    if(Term::is_stdin_a_tty()) std::cin.get();
  }
...
```

![Screenshot_20230510_010030](https://github.com/jupyter-xeus/cpp-terminal/assets/8627746/a36ebe48-c1e9-46c3-91f7-eacdf3beb452)

Right is Unifont https://unifoundry.com/unifont/, it seems free to use and to install. It is loaded programmatically, You need to know where to find the file and the name of the font inside it (maybe there is a better solution but it's working)

Anyway, as it's possible to change font in Windows it would be nice to propose this feature to the user. What do you think ? On Linux this fonction could do nothing.

**Reply by iganinja** [reply]
> [#239 (comment)](https://github.com/jupyter-xeus/cpp-terminal/pull/239#issuecomment-1539682747):
> 
> > I like Turbo Vision visual style, imitating a graphical GUI. From the programming perspective, having a `Widget` base class which can contain children and following a Qt like approach is the most familiar one to me. I did some half-baked approach on my program so far, but it would need to be more sophisticated. Should we create a new thread to talk about how to engage this development?
> 
> @iganinja, I like the Turbo Vision style as well. Married with unicode it can look really nice.
> 
> I am also familiar with the imperative Qt style. So why don't we start with that? We can always build a "react" functional style on top of this imperative style as another optional layer.

So, `Widget` base clase, parent-child relationship with `std::unique_ptr` instead of raw ones. I want to use (0, 0) based coordinate system inside Widgets, instead of (1, 1) of cpp-terminal (if I  understood it incorrectly). I have signal/slot implementation on a library of mine (not nearly as sophisticated as Qt's one, of course), do you think we need a mechanism like that?

**Reply by certik** [reply]
@flagarde nice! Can you also test this:
```c++
std::cout<<"Hebrew : עברית"<<std::endl;
```
The default terminal font always shows the Hebrew font very thin for me.

The way to do this is to ship the font in the "share/cpp-terminal" directory, and we then find this directory at runtime and load it on Windows, if the user requests it (say with `load_unifont();`). On Linux this would do nothing or give an error. And we can provide instructions how to change the font in the terminal manually for the most common terminals.

The only problem is that the Unifont font seems to be licensed under GPL2, which would I think require us to relicense cpp-terminal to GPL2 also to ship it, and I don't want to do that. So let's find another font that has a permissive license that we can ship.

@iganinja I would recommend we play with a few prototypes to see what design is the best. Yes, `std::unique_ptr` and `std::shared_ptr` is the way to go. I think terminals are traditionally numbered 1..80 by 1..25, so I would stick to that. But if you feel strongly you want 0 based, you can, just make sure to document it well that it is different than the underlying terminal coordinate system that is 1-based, per: https://en.wikipedia.org/wiki/ANSI_escape_code#CSI_(Control_Sequence_Introducer)_sequences, "Moves the cursor to row n, column m. The values are 1-based, and default to 1 (top left corner) if omitted.". Go ahead and create a simple demo, say the menu (that you already have I think) and a dialog with a button when a menu item is selected. That should show almost everything: defining widgets, callbacks, the main loop that listens to arrow keys and enter to select menu items and clicks a button. You can send it as a PR. We can look at the code and see if the design is clean and maintainable. We can iterate on it until we are happy, then we can merge and send PRs with more widgets.

**Reply by flagarde** [reply]
@certik I can try for sure. I thought this font was quite flexible... i will dig more into it.

I was thinking putting by default a path as you said and then to read an environment variable programer would set/change. This would allow to have cpp-terminal default font in its directory. Programer own fonts and use installed font if he wants too.

I would vote for : do nothing and not crash on linux. It seems most Linuxes have good font defined as default Linux users would maybe more aware about how to change their font in their terminals

I did a PR to have a doxygen generation and we would just have to tell on Unix this function do nothing and that's it :)

**Reply by certik** [reply]
@flagarde that is fine with me. Go ahead and work with @MCWertGaming to review your PRs. If he is busy, you can ping me too. I don't have typically time to review every day, but I also don't want you to be blocked, so just ping me as needed.

Thank you @flagarde and @iganinja for your work on this, I am super excited about it. Both solid utf8 support and beautiful widgets  will be awesome and we'll use it in LFortran / LPython as an optional terminal based GUI. Right now we use the multiline prompt facility, and I want to also add optional GUI with menus and widgets to naturally expose all the features that the compilers offer. Possibly even in the style of Turbo Pascal, with the builtin debugger, etc. I don't think we have to focus on the editor part, but for debugging to understand how a given code works, that would be awesome, as well as various interactive interaction with the intermediate representations that the compiler has during compilation, I really like having a good solid GUI in a terminal.

**Reply by flagarde** [reply]
@certik They say is dual license : As of Unifont version 13.0.04, the fonts are dual-licensed under the SIL Open Font License (OFL) version 1.1 and the GNU GPL 2+ with the GNU font embedding exception. The SIL OFL is available at[ OFL-1.1.txt](https://unifoundry.com/OFL-1.1.txt).

And they say : 

```
The copyright holders have chosen to grant rights of use to GNU Unifont versions, the GNU
bitmapped letter forms and the GNU icon designs under the terms of the GPL-2.0+, thus
generally giving everyone the freedom to use, study, change, access and redistribute GNU
Unifont versions, the GNU bitmapped letter forms and the GNU icon designs, as well as the
freedom to distribute modified versions of them.
However, these freedoms are granted under the condition that the requirements defined by
the GPL-2.0+ are met. This especially requires third parties distributing derivative works of
GNU Unifont versions, the GNU bitmapped letter forms or the GNU icon designs to license
these derivative works under the terms of the GPL-2.0+.
Accordingly, any attempt to license derivative works of GNU Unifont versions under other
terms or to commit them to the public domain is an infringement of the GPL-2.0’s terms
```

I'm not a lawer but to me this mean you can distribute it without change and provide the user the GPL2+ syaing this font is licensed with GPL2+ and it seems we can even embedded it (even if I don't know what they means by embedded , if we can read the file as binary and included it this way as .cpp for example). But IMHO, it seems redistributing it is fine and the license would not leaks on cpp-terminal. Maybe a lawer could help :)

**Reply by certik** [reply]
@flagarde we'll probably don't want to commit the font into this git repository, so I assume we provide the tooling for end applications (that use cpp-terminal) to easily download & ship it with the application. And then they can figure out the licensing situation and we don't have to worry about it.

**Reply by flagarde** [reply]
@certik Yes exactly, I was thinking to add a CMake option to download the font and put it in the right folder. So it would'nt be on the cpp-terminal repo and so it would be the programmer responsbility to install it or not. I think this way we are safe and we stay away from GPL-2.0+ right ?

**Reply by flagarde** [reply]
But I want to offer this opportunity especially for debugging. To use same font in Unix/Windows would be beneficial for tests I think

**Reply by certik** [reply]
@flagarde I think that should work. My understanding of GPL2 is that as long as your dependency on it is optional, that is cpp-terminal builds and works just fine without it, then we don't have any obligations: they only apply if the end user turns the optional dependency on. Yes, having some kind of a "canonical" font that we test that works on all platforms is good, we should do it.

**Reply by flagarde** [reply]
@certik Perfect so we just add an option with a #Please check licensing and that's it 👍🏻

**Comment by flagarde** [comment]
@certik With Hebrew 

![Screenshot_20230510_055801](https://github.com/jupyter-xeus/cpp-terminal/assets/8627746/020b037f-4c72-4810-92be-49991d4c3d05)

**Reply by certik** [reply]
The font looks good! But we have a bug in the renderer --- the word is flipped horizontally, compare it with the text I sent above. Otherwise it looks good.

I can't read Arabic, but it might be similar, and also it might not be connected correctly, just comparing it against the source code visually.

**Reply by flagarde** [reply]
@certik bidi is maybe not done in terminal by default, We maybe have to do this by hand. Ligatures etc can't be performed in the standard console cmd.exe I think (Windows negelect consoles until very recently). Maybe with the new one is better ?

**Comment by iganinja** [comment]
By the way, do we have [FTXUI](https://arthursonzogni.github.io/FTXUI/) on our list? It's another C++ terminal library, although in some fast tests I was not able to make it run properly in Windows.

**Reply by certik** [reply]
Yes, we had it in the README, but it got removed. Let's put it back in: https://github.com/jupyter-xeus/cpp-terminal/issues/272

**Comment by iganinja** [comment]
So, I come up with this `Widget` class, very heavily inspired in Qt's one:

```
class Widget
{
public:
  Widget(Widget* parent = nullptr);
  virtual ~Widget() = default;

  virtual Widget* add_child(std::unique_ptr<Widget>&& child);
  virtual std::unique_ptr<Widget> remove_child(Widget* child);

  virtual Position position() const;
  virtual void set_position(Position position);

  virtual Size size() const;
  virtual void resize(Size available_size);

  virtual void draw(Window& target);

  virtual bool is_visible() const;
  virtual void set_visible(bool visible);
  virtual void show();
  virtual void hide();
  
  enum class InputProcessing
  {
    Processed,
    Skipped
  };
  
  virtual InputProcessing proces_input(Key key);
  
  bool has_focus() const;
  void set_focus(bool focus);

private:
  Widget* m_parent{nullptr};
  bool m_is_visible{true};
  std::vector<std::unique_ptr<Widget>> m_children;
  bool m_has_focus{false};
};
```

Each `Widget` owns N children. You can add or remove them at any time. We have position and size getters and setters. `resize()` is somewhat a hit. For example, if we have a `Menu`, you can call its `resize()` with `(10, 20)` size, but `Menu` after that call with return `(10, 1)` size, as `Menu` will always have height of 1.

`draw()` of course draws the widget. I'm thinking we will need to rework `Window` in order to allow not visible characters, currently it throws an exception if you try to draw outside its bounds. This way widget handling will be much easier, developing my editor I saw it's very easy to draw where I should not. Anyway I'm also thinking we need to clip somehow the widgets, as if we want to implement a `ScrollView` we would need it. Perhaps passing a `Rect` to `draw()`? There is also the option to pass a class like `ClippingWindow` instead of `Window`, and this `ClippingWindow` be configured with a rect which indicates where drawing operations are allowed. 

Finally, I'm not really sure how to implement the focus thing. If, for example, we have a form with several line inputs (name, surname, etc), we would want to put the blinking caret in the first one, then press tab and set the focus to the next one. How can we implement that? What would be the correct way to unfocus one widget and focus another one?

Well, let me know what do you think about it, please. Thanks!

**Comment by iganinja** [comment]
Hi guys.

I'm doing a little bit of work on that. Can we create a branch, something like `widgets_poc` so I can upload my stuff to there?

**Reply by flagarde** [reply]
You can't create one?

**Reply by flagarde** [reply]
@iganinja I have created a branch for your work
